experimental/ssh/internal/proxy/connections_test.go

```diff
@@ -0,0 +1,257 @@
+	"sync"
+	"testing"
+	"time"
+
+	// TODO: re-enable synctests after we update to Go 1.25
+	// "testing/synctest"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestConnectionsManager_TryAdd_Success(t *testing.T) {
+	cm := NewConnectionsManager(2, time.Hour)
+	conn1 := &proxyConnection{}
+	conn2 := &proxyConnection{}
+
+	// Should succeed for first connection
+	result := cm.TryAdd("conn1", conn1)
+	assert.True(t, result)
+	assert.Equal(t, 1, cm.Count())
+
+	// Should succeed for second connection
+	result = cm.TryAdd("conn2", conn2)
+	assert.True(t, result)
+	assert.Equal(t, 2, cm.Count())
+
+	// Verify connections can be retrieved
+	retrieved1, exists1 := cm.Get("conn1")
+	assert.True(t, exists1)
+	assert.Equal(t, conn1, retrieved1)
+
+	retrieved2, exists2 := cm.Get("conn2")
+	assert.True(t, exists2)
+	assert.Equal(t, conn2, retrieved2)
+}
+
+func TestConnectionsManager_TryAdd_MaxClientsReached(t *testing.T) {
+	cm := NewConnectionsManager(1, time.Hour)
+	conn1 := &proxyConnection{}
+	conn2 := &proxyConnection{}
+
+	// First connection should succeed
+	result := cm.TryAdd("conn1", conn1)
+	assert.True(t, result)
+	assert.Equal(t, 1, cm.Count())
+
+	// Second connection should fail due to max clients limit
+	result = cm.TryAdd("conn2", conn2)
+	assert.False(t, result)
+	assert.Equal(t, 1, cm.Count())
+
+	// Only first connection should exist
+	_, exists1 := cm.Get("conn1")
+
+	assert.True(t, exists1)
+
+	_, exists2 := cm.Get("conn2")
+	assert.False(t, exists2)
+}
+
+func TestConnectionsManager_Remove(t *testing.T) {
+	cm := NewConnectionsManager(3, time.Hour)
+	conn := &proxyConnection{}
+
+	// Add connection
+	result := cm.TryAdd("test-id", conn)
+	assert.True(t, result)
+	assert.Equal(t, 1, cm.Count())
+
+	// Remove connection
+	cm.Remove("test-id")
+	assert.Equal(t, 0, cm.Count())
+
+	// Connection should no longer exist
+	_, exists := cm.Get("test-id")
+	assert.False(t, exists)
+
+	// Remove non-existent connection should not panic
+	cm.Remove("non-existent")
+	assert.Equal(t, 0, cm.Count())
+}
+
+func TestConnectionsManager_ThreadSafety(t *testing.T) {
+	cm := NewConnectionsManager(100, time.Hour)
+	const numGoroutines = 10
+	const numOperationsPerGoroutine = 100
+
+	var wg sync.WaitGroup
+
+	for i := range numGoroutines {
+		wg.Add(1)
+		go func(routineID int) {
+			defer wg.Done()
+			for j := range numOperationsPerGoroutine {
+				cm.TryAdd(fmt.Sprintf("conn-%d-%d", routineID, j), &proxyConnection{})
+			}
+		}(i)
+	}
+
+	wg.Wait()
+
+	// Verify we can get the count without race conditions
+	countAfterAdds := cm.Count()
+	assert.LessOrEqual(t, countAfterAdds, numGoroutines*numOperationsPerGoroutine)
+
+	// Now do concurrent gets, removes, and counts
+	for i := range numGoroutines {
+		wg.Add(1)
+		go func(routineID int) {
+			defer wg.Done()
+			for j := range numOperationsPerGoroutine {
+				cm.Get(fmt.Sprintf("conn-%d-%d", routineID, j))
+			}
+		}(i)
+	}
+
+	for i := range numGoroutines {
+		wg.Add(1)
+		go func(routineID int) {
```

Comment:
FWIW, you don't need to capture the loop var in an argument anymore, you can refer to it and it will be captured by value as of Go 1.22: https://go.dev/blog/loopvar-preview